File: branches/scala_implementation/taboosearch/src/taboosearch/tsp/TspTaboolator.scala
package taboosearch.tsp

import core._;
import taboosearch.tenures._
import taboosearch._;
import scala.collection.mutable.{ListBuffer, HashMap};

class TspTaboolator(val strategy : TenureStrategy) extends Taboolator[TspSolution] {
	val taboo = new HashMap[Int, Int]; // vertex -> tenure
	
	def setTabu(solution : TspSolution) = {
		val (i, j) = solution.move;
		val v1 = solution(i);
		val v2 = solution(j);
		val currentTenure = strategy.getTenure(); 
		taboo.put(v1, currentTenure);
		taboo.put(v2, currentTenure);
	}
	
	def isTabu(solution : TspSolution) : Boolean = {
		val (i, j) = solution.move;
		val v1 = solution(i);
		val v2 = solution(j);
		return taboo.contains(v1) && taboo.contains(v2);
	}
	
	def tick() = {
		strategy.tick();
		decreaseTenures();
	}
	
	private def decreaseTenures() = {
		var toErase = new ListBuffer[Int]();
		for ((attr, tenure) <- taboo) {
			if (tenure == 1)
				toErase.append(attr);
			else {
				taboo.remove(attr);
				taboo.put(attr, tenure - 1);
			}
		}
		taboo --= toErase;
	}
	
}

File: src/taboosearch/tsp/TSPTaboolator.java
package taboosearch.tsp;

import java.util.HashMap;
import java.util.Vector;

import taboosearch.tenures.TenureStrategy;
import taboosearch.Taboolator;

public class TSPTaboolator extends Taboolator<TSPSolution, TSPSwapMove> {
	private TenureStrategy strategy;
	private HashMap<TSPAttribute, Integer> taboo;
	
	public TSPTaboolator(TenureStrategy strategy) {
		this.strategy = strategy;
		this.taboo = new HashMap<TSPAttribute, Integer>();
	}
		
	@Override
	public boolean isTabu(final TSPSolution solution, final TSPSwapMove move) {
		boolean isTabu = true;
		for (TSPAttribute attribute : move.getAttributes(solution))
			isTabu &= taboo.containsKey(attribute);
		return isTabu;
	}

	@Override
	public void setTabu(final TSPSolution solution, final TSPSwapMove move) {
		for (TSPAttribute attribute : move.getAttributes(solution))
			taboo.put(attribute, strategy.getTenure());
	}
	
	public void tick() {
		strategy.tick();
		decreaseTenures();
	}
	
	public void tick(final TSPSolution solution, final TSPSwapMove move) {
		setTabu(solution, move);
		strategy.tick();
		decreaseTenures();
	}
	
	private void decreaseTenures() {
		Vector<TSPAttribute> toErase = new Vector<TSPAttribute>();

		for (TSPAttribute attribute : taboo.keySet()) {
			int tenure = taboo.get(attribute);
			if (tenure == 1) {
				toErase.add(attribute);
			} else {
				taboo.put(attribute, tenure - 1);
			}
		}
		
		for (TSPAttribute attribute : toErase)
			taboo.remove(attribute);
	}
}
